File: modules/audio-equalizer/ios/AudioEqualizerModule.swift
import ExpoModulesCore
import AVFoundation

/**
 * Native Audio Equalizer Module for iOS
 *
 * Uses AVAudioUnitEQ to provide audio equalization through the AVAudioEngine.
 * Note: This implementation is more complex on iOS because the audio engine
 * needs to be integrated with the existing audio playback system.
 */
public class AudioEqualizerModule: Module {
    private var audioEngine: AVAudioEngine?
    private var equalizer: AVAudioUnitEQ?
    private var isInitialized = false
    private var isEnabled = false
    private var bandLevels: [Float] = []

    // Standard 10-band EQ frequencies
    private let standardFrequencies: [Float] = [32, 64, 125, 250, 500, 1000, 2000, 4000, 8000, 16000]
    private let numberOfBands = 10
    private let minBandLevel: Float = -1500  // -15 dB in millibels
    private let maxBandLevel: Float = 1500   // +15 dB in millibels

    public func definition() -> ModuleDefinition {
        Name("AudioEqualizer")

        AsyncFunction("getEqualizerInfo") { (promise: Promise) in
            DispatchQueue.global(qos: .userInitiated).async {
                do {
                    try self.initializeIfNeeded()
                    let result = self.getEqualizerInfoInternal()
                    promise.resolve(result)
                } catch {
                    promise.resolve(self.createUnavailableInfo())
                }
            }
        }

        AsyncFunction("getPresets") { (promise: Promise) in
            DispatchQueue.global(qos: .userInitiated).async {
                // iOS doesn't have built-in EQ presets like Android
                // Return empty array - presets are managed on the JS side
                promise.resolve([[String: Any]]())
            }
        }

        AsyncFunction("getState") { (promise: Promise) in
            DispatchQueue.global(qos: .userInitiated).async {
                do {
                    try self.initializeIfNeeded()
                    let result = self.getStateInternal()
                    promise.resolve(result)
                } catch {
                    promise.resolve(self.createDisabledState())
                }
            }
        }

        AsyncFunction("setEnabled") { (enabled: Bool, promise: Promise) in
            DispatchQueue.global(qos: .userInitiated).async {
                do {
                    try self.initializeIfNeeded()
                    try self.setEnabledInternal(enabled: enabled)
                    promise.resolve(nil)
                } catch {
                    promise.reject("EQ_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("setBandLevel") { (bandIndex: Int, level: Int, promise: Promise) in
            DispatchQueue.global(qos: .userInitiated).async {
                do {
                    try self.initializeIfNeeded()
                    try self.setBandLevelInternal(bandIndex: bandIndex, level: Float(level))
                    promise.resolve(nil)
                } catch {
                    promise.reject("EQ_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("setBandLevels") { (levels: [Int], promise: Promise) in
            DispatchQueue.global(qos: .userInitiated).async {
                do {
                    try self.initializeIfNeeded()
                    try self.setBandLevelsInternal(levels: levels.map { Float($0) })
                    promise.resolve(nil)
                } catch {
                    promise.reject("EQ_ERROR", error.localizedDescription)
                }
            }
        }

        AsyncFunction("usePreset") { (presetName: String, promise: Promise) in
            DispatchQueue.global(qos: .userInitiated).async {
                // iOS doesn't have built-in presets, so we just acknowledge
                // The JS side should call setBandLevels with the preset values
                promise.resolve(nil)
            }
        }

        AsyncFunction("release") { (promise: Promise) in
            DispatchQueue.global(qos: .userInitiated).async {
                self.releaseInternal()
                promise.resolve(nil)
            }
        }

        OnDestroy {
            self.releaseInternal()
        }
    }

    private func initializeIfNeeded() throws {
        guard !isInitialized else { return }

        // Initialize band levels to flat (0 gain)
        bandLevels = Array(repeating: 0, count: numberOfBands)

        // Note: Full AVAudioEngine integration would require modifying
        // the audio playback chain. For now, we store settings that can
        // be applied when integrated with the player.

        // Create the EQ unit for when we can integrate it
        let eq = AVAudioUnitEQ(numberOfBands: numberOfBands)
        equalizer = eq

        // Configure each band as a parametric EQ
        for i in 0..<numberOfBands {
            let band = eq.bands[i]
            band.filterType = .parametric
            band.frequency = standardFrequencies[i]
            band.bandwidth = 1.0 // Octave bandwidth
            band.gain = 0 // Start flat
            band.bypass = false
        }

        isInitialized = true
    }

    private func getEqualizerInfoInternal() -> [String: Any] {
        let bands: [[String: Any]] = standardFrequencies.enumerated().map { index, freq in
            // Approximate frequency range (half octave below and above)
            let minFreq = freq / 1.414 // sqrt(2)
            let maxFreq = freq * 1.414
            return [
                "minFrequency": Int(minFreq),
                "maxFrequency": Int(maxFreq),
                "centerFrequency": Int(freq)
            ]
        }

        return [
            "isAvailable": true,
            "numberOfBands": numberOfBands,
            "minBandLevel": Int(minBandLevel),
            "maxBandLevel": Int(maxBandLevel),
            "bands": bands
        ]
    }

    private func getStateInternal() -> [String: Any?] {
        return [
            "isEnabled": isEnabled,
            "bandLevels": bandLevels.map { Int($0) },
            "presetName": nil
        ]
    }

    private func setEnabledInternal(enabled: Bool) throws {
        isEnabled = enabled

        guard let eq = equalizer else { return }

        // Bypass all bands when disabled
        for band in eq.bands {
            band.bypass = !enabled
        }
    }

    private func setBandLevelInternal(bandIndex: Int, level: Float) throws {
        guard bandIndex >= 0 && bandIndex < numberOfBands else {
            throw NSError(domain: "AudioEqualizer", code: 1, userInfo: [
                NSLocalizedDescriptionKey: "Band index \(bandIndex) out of range (0-\(numberOfBands - 1))"
            ])
        }

        let clampedLevel = min(max(level, minBandLevel), maxBandLevel)
        bandLevels[bandIndex] = clampedLevel

        // Convert millibels to decibels for AVAudioUnitEQ
        let gainDb = clampedLevel / 100.0

        if let eq = equalizer {
            eq.bands[bandIndex].gain = gainDb
        }
    }

    private func setBandLevelsInternal(levels: [Float]) throws {
        for i in 0..<min(levels.count, numberOfBands) {
            try setBandLevelInternal(bandIndex: i, level: levels[i])
        }
    }

    private func releaseInternal() {
        if let engine = audioEngine, engine.isRunning {
            engine.stop()
        }
        audioEngine = nil
        equalizer = nil
        isInitialized = false
        isEnabled = false
        bandLevels = []
    }

    private func createUnavailableInfo() -> [String: Any] {
        return [
            "isAvailable": false,
            "numberOfBands": 0,
            "minBandLevel": 0,
            "maxBandLevel": 0,
            "bands": [[String: Any]]()
        ]
    }

    private func createDisabledState() -> [String: Any?] {
        return [
            "isEnabled": false,
            "bandLevels": [Int](),
            "presetName": nil
        ]
    }
}

File: modules/audio-equalizer/android/src/main/java/expo/modules/audioequalizer/AudioEqualizerModule.kt
package expo.modules.audioequalizer

import android.content.Context
import android.media.AudioManager
import android.media.audiofx.Equalizer
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

/**
 * Native Audio Equalizer Module for Android
 *
 * Uses android.media.audiofx.Equalizer to provide system-level audio equalization.
 * Attaches to the audio session used by the media player.
 */
class AudioEqualizerModule : Module() {
    private var equalizer: Equalizer? = null
    private var isInitialized = false

    private val context: Context
        get() = appContext.reactContext ?: throw IllegalStateException("React context is null")

    override fun definition() = ModuleDefinition {
        Name("AudioEqualizer")

        AsyncFunction("getEqualizerInfo") { promise: Promise ->
            runOnBackgroundThread {
                try {
                    initializeIfNeeded()
                    val result = getEqualizerInfoInternal()
                    promise.resolve(result)
                } catch (e: Exception) {
                    promise.resolve(createUnavailableInfo())
                }
            }
        }

        AsyncFunction("getPresets") { promise: Promise ->
            runOnBackgroundThread {
                try {
                    initializeIfNeeded()
                    val result = getPresetsInternal()
                    promise.resolve(result)
                } catch (e: Exception) {
                    promise.resolve(emptyList<Map<String, Any>>())
                }
            }
        }

        AsyncFunction("getState") { promise: Promise ->
            runOnBackgroundThread {
                try {
                    initializeIfNeeded()
                    val result = getStateInternal()
                    promise.resolve(result)
                } catch (e: Exception) {
                    promise.resolve(createDisabledState())
                }
            }
        }

        AsyncFunction("setEnabled") { enabled: Boolean, promise: Promise ->
            runOnBackgroundThread {
                try {
                    initializeIfNeeded()
                    setEnabledInternal(enabled)
                    promise.resolve(null)
                } catch (e: Exception) {
                    promise.reject("EQ_ERROR", e.message ?: "Failed to set enabled state", e)
                }
            }
        }

        AsyncFunction("setBandLevel") { bandIndex: Int, level: Int, promise: Promise ->
            runOnBackgroundThread {
                try {
                    initializeIfNeeded()
                    setBandLevelInternal(bandIndex, level)
                    promise.resolve(null)
                } catch (e: Exception) {
                    promise.reject("EQ_ERROR", e.message ?: "Failed to set band level", e)
                }
            }
        }

        AsyncFunction("setBandLevels") { levels: List<Int>, promise: Promise ->
            runOnBackgroundThread {
                try {
                    initializeIfNeeded()
                    setBandLevelsInternal(levels)
                    promise.resolve(null)
                } catch (e: Exception) {
                    promise.reject("EQ_ERROR", e.message ?: "Failed to set band levels", e)
                }
            }
        }

        AsyncFunction("usePreset") { presetName: String, promise: Promise ->
            runOnBackgroundThread {
                try {
                    initializeIfNeeded()
                    usePresetInternal(presetName)
                    promise.resolve(null)
                } catch (e: Exception) {
                    promise.reject("EQ_ERROR", e.message ?: "Failed to use preset", e)
                }
            }
        }

        AsyncFunction("release") { promise: Promise ->
            runOnBackgroundThread {
                try {
                    releaseInternal()
                    promise.resolve(null)
                } catch (e: Exception) {
                    promise.reject("EQ_ERROR", e.message ?: "Failed to release equalizer", e)
                }
            }
        }

        OnDestroy {
            releaseInternal()
        }
    }

    private fun runOnBackgroundThread(action: () -> Unit) {
        Thread {
            action()
        }.start()
    }

    @Synchronized
    private fun initializeIfNeeded() {
        if (isInitialized && equalizer != null) {
            return
        }

        try {
            // Use audio session ID 0 for global output mix
            // This allows the equalizer to affect all audio output
            equalizer = Equalizer(0, 0)
            isInitialized = true
        } catch (e: Exception) {
            throw RuntimeException("Failed to initialize equalizer: ${e.message}", e)
        }
    }

    private fun getEqualizerInfoInternal(): Map<String, Any> {
        val eq = equalizer ?: return createUnavailableInfo()

        val numberOfBands = eq.numberOfBands.toInt()
        val levelRange = eq.bandLevelRange
        val minLevel = levelRange[0].toInt()
        val maxLevel = levelRange[1].toInt()

        val bands = mutableListOf<Map<String, Any>>()
        for (i in 0 until numberOfBands) {
            val freqRange = eq.getBandFreqRange(i.toShort())
            val centerFreq = eq.getCenterFreq(i.toShort())
            bands.add(mapOf(
                "minFrequency" to freqRange[0],
                "maxFrequency" to freqRange[1],
                "centerFrequency" to centerFreq
            ))
        }

        return mapOf(
            "isAvailable" to true,
            "numberOfBands" to numberOfBands,
            "minBandLevel" to minLevel,
            "maxBandLevel" to maxLevel,
            "bands" to bands
        )
    }

    private fun getPresetsInternal(): List<Map<String, Any>> {
        val eq = equalizer ?: return emptyList()

        val presets = mutableListOf<Map<String, Any>>()
        val numPresets = eq.numberOfPresets.toInt()

        for (i in 0 until numPresets) {
            val presetName = eq.getPresetName(i.toShort())
            presets.add(mapOf(
                "id" to i.toString(),
                "name" to presetName
            ))
        }

        return presets
    }

    private fun getStateInternal(): Map<String, Any?> {
        val eq = equalizer ?: return createDisabledState()

        val numberOfBands = eq.numberOfBands.toInt()
        val bandLevels = mutableListOf<Int>()

        for (i in 0 until numberOfBands) {
            bandLevels.add(eq.getBandLevel(i.toShort()).toInt())
        }

        val currentPreset = try {
            val presetIndex = eq.currentPreset.toInt()
            if (presetIndex >= 0 && presetIndex < eq.numberOfPresets) {
                eq.getPresetName(presetIndex.toShort())
            } else {
                null
            }
        } catch (e: Exception) {
            null
        }

        return mapOf(
            "isEnabled" to eq.enabled,
            "bandLevels" to bandLevels,
            "presetName" to currentPreset
        )
    }

    private fun setEnabledInternal(enabled: Boolean) {
        val eq = equalizer ?: throw IllegalStateException("Equalizer not initialized")
        eq.enabled = enabled
    }

    private fun setBandLevelInternal(bandIndex: Int, level: Int) {
        val eq = equalizer ?: throw IllegalStateException("Equalizer not initialized")

        val numberOfBands = eq.numberOfBands.toInt()
        if (bandIndex < 0 || bandIndex >= numberOfBands) {
            throw IllegalArgumentException("Band index $bandIndex out of range (0-${numberOfBands - 1})")
        }

        val levelRange = eq.bandLevelRange
        val clampedLevel = level.coerceIn(levelRange[0].toInt(), levelRange[1].toInt())
        eq.setBandLevel(bandIndex.toShort(), clampedLevel.toShort())
    }

    private fun setBandLevelsInternal(levels: List<Int>) {
        val eq = equalizer ?: throw IllegalStateException("Equalizer not initialized")

        val numberOfBands = eq.numberOfBands.toInt()
        val levelRange = eq.bandLevelRange

        for (i in 0 until minOf(levels.size, numberOfBands)) {
            val clampedLevel = levels[i].coerceIn(levelRange[0].toInt(), levelRange[1].toInt())
            eq.setBandLevel(i.toShort(), clampedLevel.toShort())
        }
    }

    private fun usePresetInternal(presetName: String) {
        val eq = equalizer ?: throw IllegalStateException("Equalizer not initialized")

        val numPresets = eq.numberOfPresets.toInt()
        for (i in 0 until numPresets) {
            if (eq.getPresetName(i.toShort()) == presetName) {
                eq.usePreset(i.toShort())
                return
            }
        }

        throw IllegalArgumentException("Preset '$presetName' not found")
    }

    @Synchronized
    private fun releaseInternal() {
        try {
            equalizer?.release()
        } catch (e: Exception) {
            // Ignore release errors
        }
        equalizer = null
        isInitialized = false
    }

    private fun createUnavailableInfo(): Map<String, Any> {
        return mapOf(
            "isAvailable" to false,
            "numberOfBands" to 0,
            "minBandLevel" to 0,
            "maxBandLevel" to 0,
            "bands" to emptyList<Map<String, Any>>()
        )
    }

    private fun createDisabledState(): Map<String, Any?> {
        return mapOf(
            "isEnabled" to false,
            "bandLevels" to emptyList<Int>(),
            "presetName" to null
        )
    }
}
